extensions/amp-share-tracking/0.1/amp-share-tracking.js

```diff
@@ -15,43 +15,115 @@
+    user.assert(this.isExperimentOn_(), `${TAG} experiment is disabled`);
+
+    /** @private {string} */
+    this.vendorHref_ = this.element.getAttribute('data-href');
+    dev.fine(TAG, 'vendorHref_: ', this.vendorHref_);
+
+    /** @private {!Promise<!Object>} */
+    this.shareTrackingFragments_ = Promise.all([
+      this.getIncomingFragment_(),
+      this.getOutgoingFragment_()]).then(results => {
+        dev.fine(TAG, 'incomingFragment: ', results[0]);
+        dev.fine(TAG, 'outgoingFragment: ', results[1]);
+        return {
+          incomingFragment: results[0],
+          outgoingFragment: results[1],
+        };
+      });
+  }
+
+  /**
+   * Get the incoming share-tracking fragment from the viewer
+   * @return {!Promise<string|undefined>}
+   * @private
+   */
+  getIncomingFragment_() {
+    return viewerFor(this.win).getFragment().then(fragment => {
+      if (!fragment || fragment.indexOf('.') != 0) {
+        return;
+      }
+      const endIndex = fragment.indexOf('&');
+      if (endIndex != -1) {
+        return fragment.substr(1, (endIndex - 1));
+      } else {
+        return fragment.substr(1);
+      }
+    });
+  }
+
+  /**
+   * Get an outgoing share-tracking fragment
+   * @return {!Promise<string|undefined>}
+   * @private
+   */
+  getOutgoingFragment_() {
+    if (this.vendorHref_) {
+      return this.getOutgoingFragmentFromVendor_(this.vendorHref_);
     }
+    return this.getOutgoingRandomFragment_();
   }
-}
 
-/**
- * ShareTrackingService processes the incoming and outcoming url fragment
- */
-export class ShareTrackingService {
-  constructor(window) {
-    this.win = window;
+  /**
+   * Get an outgoing share-tracking fragment from vendor
+   * by issueing a post request to the url the vendor provided
+   * @param {string} vendorUrl
+   * @return {!Promise<string|undefined>}
+   * @private
+   */
+  getOutgoingFragmentFromVendor_(vendorUrl) {
+    const postReq = {
+      method: 'POST',
+      credentials: 'include',
+      requireAmpResponseSourceOrigin: true,
+      body: {},
+    };
+    return xhrFor(this.win).fetchJson(vendorUrl, postReq).then(response => {
+      if (response.fragment) {
+        return response.fragment;
+      }
+      user.error(TAG, 'The response from [' + vendorUrl + '] does not ' +
+          'have a fragment value.');
+    }).catch(error => {
+      user.error(TAG, 'The request to share-tracking endpoint failed:' + error);
+    });
   }
-}
 
-/**
- * @param {!Window} window
- * @return {!ShareTrackingService}
- */
-export function installShareTrackingService(window) {
-  return getService(window, 'shareTrackingService', () => {
-    return new ShareTrackingService(window);
-  });
+  /**
+   * Get a random outgoing share-tracking fragment
+   * @return {!Promise<string>}
+   * @private
+   */
+  getOutgoingRandomFragment_() {
+    // TODO(yuxichen): Generate random outgoing fragment
+    const randomFragment = 'rAmDoM';
+    return Promise.resolve(randomFragment);
```

Comment:
I prefer making the return type of `getOutgoingFragment_` simple, otherwise I would make the method that calls it confused.